sdk/include/opentelemetry/sdk/logs/logger_provider.h

```diff
@@ -0,0 +1,102 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include "opentelemetry/logs/logger_provider.h"
+#include "opentelemetry/logs/noop.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/sdk/common/atomic_shared_ptr.h"
+#include "opentelemetry/sdk/logs/logger.h"
+#include "opentelemetry/sdk/logs/processor.h"
+
+// Define the maximum number of loggers that are allowed to be registered to the loggerprovider.
+// TODO: Add link to logging spec once this is added to it
+#define MAX_LOGGER_COUNT 100
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace logs
+{
+class LoggerProvider final : public opentelemetry::logs::LoggerProvider
+{
+public:
+  /**
+   * Initialize a new logger provider with a specified processor
+   * @param processor The processor for this logger provider. This must
+   * not be a nullptr.
+   */
+  explicit LoggerProvider(std::shared_ptr<LogProcessor> processor) noexcept;
+
+  /**
+   * Initialize a new logger provider. A processor must later be assigned
+   * to this logger provider via the SetProcessor() method.
+   */
+  explicit LoggerProvider() noexcept;
+
+  /**
+   * Creates a logger with the given name, and returns a shared pointer to it.
+   * If a logger with that name already exists, return a shared pointer to it
+   * @param name The name of the logger to be created.
+   * @param options (OPTIONAL) The options for the logger. TODO: Give list of options
+   */
+  opentelemetry::nostd::shared_ptr<opentelemetry::logs::Logger> GetLogger(
+      opentelemetry::nostd::string_view name,
+      opentelemetry::nostd::string_view options = "") noexcept override;
+
+  /**
+   * Creates a logger with the given name, and returns a shared pointer to it.
+   * If a logger with that name already exists, return a shared pointer to it
+   * @param name The name of the logger to be created.
+   * @param args (OPTIONAL) The arguments for the logger.
+   */
+  opentelemetry::nostd::shared_ptr<opentelemetry::logs::Logger> GetLogger(
+      opentelemetry::nostd::string_view name,
+      nostd::span<nostd::string_view> args) noexcept override;
+
+  /**
+   * Returns a shared pointer to the processor currently stored in the
+   * logger provider. If no processor exists, returns a nullptr
+   */
+  std::shared_ptr<LogProcessor> GetProcessor() noexcept;
+
+  // Sets the common processor that all the Logger instances will use
```

Comment:
Question on these docs:

Does setting a new processor update ALL Logger instances that have already been created, or does it only change Logger instances that were created by this instance of LoggerProvider.

Additionally, is it worth denoting concurrency behavior here?

(i.e. this will lock acquiring new Loggers, but won't stop logs from being written due to the atomic-swap on the pointer).

Reply:
Hey Josh, that's a good question. Currently when the SetProcessor() method is called, it will change all logger instances that were created by this LoggerProvider to use this new processor. However, since the LoggerProvider is expected to be a singleton (like what was done in [metrics](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#obtaining-a-meter)) we don't have to worry about having different LoggerProviders making logger instances. 

This `SetProcessor(p)` functionality isn't solidified at all, and it will most likely be removed in a future PR. For example, if we'd like each logger instance to have a different processor, having a common processor defined in the LoggerProvider doesn't make much sense! Because of this future removal, we didn't focus on the concurrency issues that arise from this method.

Hope that answers your questions!

Reply:
I think we probably would want this:
1. Logger will have a pointer to the LoggerProvider. Logger should not own processor.
2. LoggerProvider will own the processor(s).
3. Logger should always respect the current processor(s) on the LoggerProvider.
4. LoggerProvider can allow AddProcessor at any time, and it should take effect on all the existing loggers PLUS future loggers created from this provider.

Reply:
Sure I can make those changes, but I have a slight modification that could be pretty useful. 

The LoggerProvider would store all the processors that the loggers would use, but it would be stored inside a map with the name of the processor as the key. So for example, if there were two processors that export to different destinations, it could look like:
```
{
  {"ConsoleProcessor", shared_ptr<Processor>}
  {"ElasticSearchProcessor", shared_ptr<Processor>}
}
```

Then, when a logger is created via the GetLogger(), the user could specify which processor it will use by passing a second argument. So instead of `GetLogger(name)`, they would call `GetLogger(name, processor_name)`. By doing it this way, different loggers can have different processors.

If this way just complicates things then I can do it the way Reiley mentioned above.

Reply:
I am very confused. There is no console processor or ElasticSearch processor. These should be exporters.
Processors have strict ordering.

Reply:
What I meant was that each processor has exactly one exporter attached to it, so the processor will indirectly have an export destination. So there could be, for example, a batch processor with a console exporter and another batch exporter with an ElasticSearch exporter.

Reply:
I think in future we can use some outside configurator class to map named logger to exporter, how this is done in log4j... I think I have seen a notion of configurator in some other OpenTelemetry SDK (python?)

Reply:
Yeah, my expectation (from other Logging API designs) is that the `Logger` class itself is just there to attach some semantics (the name of it) and fire log-events into a buffer for processing (the log-processor).  All routing of logs is done centrally with config.

We *don't* have to do that in OpenTelemetry, but if we don't I'd love to better understand why and what trade-off you're looking for.

Regarding singleton vs. separate instances of LoggerProvider, I think we should try to honor the tracing spec here w/ [TracerProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracerprovider).  The TL;DR; 

- The LogProcessor "singleton" is used by default.  This allows APIs to add telemetry information that can be routed *later* when an Application-Developer wires + configured OT.    The reason this is needed is because the API-developer doesn't know where the telemetry (logs) SHOULD go, they just want to issue them.
- A second uses case of LogsProcessor is for applications where there is a strict need for separation of Telemetry (i.e. mutli-tenant).  In this case the application is expected to instantiate and share their own LogProcessors to ensure strict separation.